docs/rfcs/0041-progressive-rollout.md

```diff
@@ -0,0 +1,296 @@
+* The interval of time for each scheduled time must be at least 5 minutes.
+* Users can stop Progressive Rollout temporary.
+* Users can use both Progressive Rollout and Feature Flag Trigger at the same time.
+* The operation of Progressive Rollout is done regardless of whether the feature flag is enabled or disabled.
+
+# Processing flow
+
+The following image is a processing flow in this feature.
+
+![processing-flow](./images/0041-image3.png)
+
+For instance, Web Client sends clause as follows:
+
+```go
+&autoopsproto.ProgressiveRolloutClause{
+	// The another varition id is vid-2
+	VariationId: "vid-1",
+	Schedules: []*autoopsproto.ProgressiveRolloutClause_Schedule{
+		{
+			// '2023-01-01 00:03:00'
+			Time: 1672498980,
+			Weight: 20000,
+		},
+		{
+			// '2023-01-01 00:06:00'
+			Time: 1672499160,
+			Weight: 40000,
+		},
+		{
+			// '2023-01-01 00:09:00'
+			Time: 1672499340,
+			Weight: 60000,
+		},
+	},
+}
+```
+
+1. Web Client registers the above clauses by calling `CreateProgressiveRollout` rules
+2. Batch service calls `ListProgressiveRollout`, and check if the current time is a scheduled time. In this case, it checks whether the current time is 2023-01-01 00:03:00.
+3. If the current time is a scheduled time and the rule is enabled, Batch service call `ExecuteProgressiveRollout`.
+4. AutoOps service calls `UpdateFeatureTargeting` to update feature rules. In this case, update the weight of vid-1 to 20000 and the weight of vid-2 to 80000.
+
+# Changes
+
+## Table
+
+We'll create `ops_progressive_rollout` table as follows. `ProgressiveRolloutManualScheduleClause` and `ProgressiveRolloutAutomaticScheduleClause` are converted into Any type and stored into `clause` column.
+
+```sql
+CREATE TABLE IF NOT EXISTS `ops_progressive_rollout` (
+  `id` VARCHAR(255) NOT NULL,
+  `feature_id` VARCHAR(255) NOT NULL,
+  `clause` JSON NOT NULL,
+  `status` INT(11) NOT NULL,
+  `created_at` BIGINT(20) NOT NULL,
+  `updated_at` BIGINT(20) NOT NULL,
+  `disabled` TINYINT(1) NOT NULL DEFAULT '0',
+  `environment_namespace` VARCHAR(255) NOT NULL,
+  PRIMARY KEY (`id`, `environment_namespace`),
+  CONSTRAINT `foreign_progressive_rollout_feature_id_environment_namespace`
+    FOREIGN KEY (`feature_id`, `environment_namespace`)
+    REFERENCES `feature` (`id`, `environment_namespace`)
+    ON DELETE NO ACTION
+    ON UPDATE NO ACTION
+);
+```
+
+## Proto
+
+### Progressive rollout
+
+* proto/autoops/progressive_rollout.proto
+
+`Status` represents the state of operations for ProgressiveRollout. If a operation has not started, `Status` is WAITING.
+If a operation is in progress, `Status` is DOING. If a operaiton is finished, `Status` is DONE.
+All fields in `ProgressiveRollout` are stored to DB as columns.
+The reason of using `google.protobuf.Any` is for ease of expansion.
+
+```proto
+message ProgressiveRollout {
```

Comment:
I was talking to the Designer about this feature today.
If we add the Type property in the proto and in the DB, we will have more flexibility to create the design making it easier to develop without checking the clause type in many places.
So let's go with the Type property as you suggested before.

E.g.
```proto
message ProgressiveRollout {
  enum Type {
    UNKNOWN = 0;
    MANUAL = 1;
    TEMPLATE = 2;
  }
  enum Status {
    UNKNOWN = 0;
    WAITING = 1;
    RUNNING = 2;
    FINISHED = 3;
    DISABLED = 4;
  }
  string id = 1;
  string feature_id = 2;
  Type type = 3;
  google.protobuf.Any clause = 4;
  Status status = 5;
  int64 created_at = 6;
  int64 updated_at = 7;
}
```